x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/CheckLicenseTests.java

```diff
@@ -0,0 +1,138 @@
+                : LicensedFeature.persistent("test", "license", functionLicense);
+            final EsqlFunctionRegistry.FunctionBuilder builder = (source, expression, cfg) -> {
+                final LicensedFunction licensedFunction = new LicensedFunction(source);
+                licensedFunction.setLicensedFeature(functionLicenseFeature);
+                return licensedFunction;
+            };
+            for (License.OperationMode operationMode : License.OperationMode.values()) {
+                if (License.OperationMode.TRIAL != operationMode && License.OperationMode.compare(operationMode, functionLicense) < 0) {
+                    // non-compliant license
+                    final VerificationException ex = expectThrows(VerificationException.class, () -> analyze(builder, operationMode));
+                    assertThat(ex.getMessage(), containsString("current license is non-compliant for function [license()]"));
+                } else {
+                    // compliant license
+                    assertNotNull(analyze(builder, operationMode));
+                }
+            }
+        }
+    }
+
+    private LogicalPlan analyze(EsqlFunctionRegistry.FunctionBuilder builder, License.OperationMode operationMode) {
+        final FunctionDefinition def = EsqlFunctionRegistry.def(LicensedFunction.class, builder, "license");
+        final EsqlFunctionRegistry registry = new EsqlFunctionRegistry(def) {
+            @Override
+            public EsqlFunctionRegistry snapshotRegistry() {
+                return this;
+            }
+        };
+        return analyzer(registry, operationMode).analyze(parser.createStatement(esql));
+    }
+
+    private static Analyzer analyzer(EsqlFunctionRegistry registry, License.OperationMode operationMode) {
+        return new Analyzer(
+            new AnalyzerContext(EsqlTestUtils.TEST_CFG, registry, analyzerDefaultMapping(), defaultEnrichResolution()),
+            new Verifier(new Metrics(new EsqlFunctionRegistry()), getLicenseState(operationMode))
+        );
+    }
+
+    private static XPackLicenseState getLicenseState(License.OperationMode operationMode) {
+        final TestUtils.UpdatableLicenseState licenseState = new TestUtils.UpdatableLicenseState();
+        licenseState.update(new XPackLicenseStatus(operationMode, true, null));
+        return licenseState;
+    }
+
+    // It needs to be public because we run validation on it via reflection in org.elasticsearch.xpack.esql.tree.EsqlNodeSubclassTests.
+    // This test prevents to add the license as constructor parameter too.
+    public static class LicensedFunction extends Function {
+
+        private LicensedFeature licensedFeature;
+
+        public LicensedFunction(Source source) {
+            super(source, List.of());
+        }
+
+        void setLicensedFeature(LicensedFeature licensedFeature) {
+            this.licensedFeature = licensedFeature;
```

Comment:
In production code how will we get the licensed feature into the functions? Perhaps they can be hard-coded references to static license features?

Reply:
yes